### PR TITLE
[checkpointio] Unsharded Optimizer Checkpoint for Gemini Plugin

### PR DESCRIPTION
**Pull request by Fridge003**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

design: #4140

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

- Implement the unsharded optimizer checkpointing feature through overloading state_dict()/load_state_dict() methods for ZeroOptimizer.
- Support compatibility with pytorch model/optimizer checkpoints and add relevant tests.

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/booster/plugin/gemini_plugin.py [hunk 1]
```diff
@@ -117,6 +119,30 @@ def load_sharded_model(self,
...
+
+    def load_sharded_optimizer(self, optimizer: Optimizer, checkpoint_index_file: Path, prefix: str):
+        """
+        Loading sharded optimizer from checkpoint folder, with index file given.
+        For each process, only loading optimizer states of parameters it controls.
+        """
+        # TODO(Baizhou): To be implemented.
+        pass
+
+    def save_lr_scheduler(self, lr_scheduler: LRScheduler, checkpoint: str):
```
Comment (ver217) [inline colossalai/booster/plugin/gemini_plugin.py:139]: Is this method different from it's base class's? If not, we don't need to rewrite it.

File: colossalai/booster/plugin/gemini_plugin.py [hunk 2]
```diff
@@ -219,7 +245,7 @@ def __init__(
         min_chunk_size_m: float = 32,
         memstats: Optional[MemStats] = None,
         gpu_margin_mem_ratio: float = 0.0,
-        initial_scale: float = 2**32,
+        initial_scale: float = 2**16,
```
Comment (ver217) [inline colossalai/booster/plugin/gemini_plugin.py:248]: You should also update docstring.

File: colossalai/checkpoint_io/utils.py [hunk 3]
```diff
@@ -88,6 +90,17 @@ def is_safetensor_checkpoint(checkpoint_file_path: str) -> bool:
 # ======================================
 # Helper functions for saving shard file
 # ======================================
+def unwrap_optimizer(optimizer: OptimizerWrapper):
+    '''
+    Unwrap a wrapped optimizer.
+    This method should be used before saving/loading it to/from sharded checkpoints.
+    '''
```
Comment (ver217) [inline colossalai/checkpoint_io/utils.py:97]: Please add a TODO comment here. `ColossalaiOptimizer` will be replaced with `OptimizerWrapper` in the future.

File: colossalai/booster/plugin/gemini_plugin.py (hunk 1 above)
Comment (Fridge003) [reply colossalai/booster/plugin/gemini_plugin.py:139]: Yes, they are the same.

File: colossalai/booster/plugin/gemini_plugin.py (hunk 2 above)
Comment (Fridge003) [reply colossalai/booster/plugin/gemini_plugin.py:248]: Updated

File: colossalai/checkpoint_io/utils.py (hunk 3 above)
Comment (Fridge003) [reply colossalai/checkpoint_io/utils.py:97]: Updated

File: colossalai/testing/comparison.py [hunk 4]
```diff
@@ -33,25 +38,35 @@ def assert_equal_in_group(tensor: Tensor, process_group: ProcessGroup = None):
 
 
 def check_state_dict_equal(d1: OrderedDict, d2: OrderedDict, ignore_device: bool = True):
-    for k, v in d1.items():
-        if isinstance(v, dict):
-            check_state_dict_equal(v, d2[k])
-        elif isinstance(v, list):
-            for i in range(len(v)):
-                if isinstance(v[i], torch.Tensor):
+    assert len(list(d1.keys())) == len(list(d2.keys()))
```
[the remark this answers is not in the document]
Comment (Fridge003) [reply colossalai/testing/comparison.py:41]: After changing to tree_flatten version of comparison, I found that some former tests will fail. Our demand of equal state_dict might be looser than that of equal flattened pytree (pytree asks for a strict order of elements), so I will keep the former method for checking equality of state_dict.

[the remark this answers is not in the document]
Comment (ver217) [reply colossalai/testing/comparison.py:42]: ok

File: colossalai/zero/gemini/gemini_optimizer.py [hunk 5]
```diff
@@ -301,25 +308,365 @@ def get_range_pair(local_chunk: Chunk, local_param: Parameter):
...
+        #                             relative to the whole optimizer state.
+        #          param_size(int): Length of parameter shard owned by current process.
+
+        assert param_id in self.id_to_fake_params
+        fake_param = self.id_to_fake_params[param_id]
+        chunk = self.param_to_chunk32[fake_param].paired_chunk
+        param = self.id_to_real_params[param_id]
+        param_info = chunk.tensors_info[param]
+
+        param_range = self.param_to_range[fake_param]
```
Comment (ver217) [inline colossalai/zero/gemini/gemini_optimizer.py:356]: `begin_in_chunk, end_in_chunk = self.param_to_range[fake_param]` is simpler

File: colossalai/zero/gemini/gemini_optimizer.py [hunk 6]
```diff
@@ -301,25 +308,365 @@ def get_range_pair(local_chunk: Chunk, local_param: Parameter):
...
                 fake_params_list.append(fake_param)
 
+            # Update self.optim.param_groups as well as backup group.
             group['params'] = fake_params_list
+            group_backup['params'] = group_ids
+            self.param_groups_backup.append(group_backup)
+
+    def get_offsets(self, param_id: int) -> tuple:
+        # Args: param_id(int): The id of parameter.
+        #
```
Comment (ver217) [inline colossalai/zero/gemini/gemini_optimizer.py:344]: The format of docstring is wrong

File: colossalai/zero/gemini/gemini_optimizer.py [hunk 7]
```diff
@@ -301,25 +308,365 @@ def get_range_pair(local_chunk: Chunk, local_param: Parameter):
...
+        collected_states = {}
+
+        # If the chunk is kept gathered,
+        # the parameteres are treated the same as that of those in strict DDP during training.
+        # So states can be directly fetched from current device.
+        if chunk.keep_gathered:
+            assert param_id in self.id_to_fake_params
+            fake_param = self.id_to_fake_params[param_id]
+            if rank == master_rank:
+                states = self.optim.state[fake_param]
```
Comment (ver217) [inline colossalai/zero/gemini/gemini_optimizer.py:394]: This `states` is a dict whose key is `state_name`. Can we just get state name through it directly?

File: colossalai/zero/gemini/gemini_optimizer.py [hunk 8]
```diff
@@ -301,25 +308,365 @@ def get_range_pair(local_chunk: Chunk, local_param: Parameter):
...
+                num_states = len(state_names)
+                container_size = 1 + (num_states - 1) * shard_size if 'step' in state_names \
+                                        else num_states * shard_size
+                container_tensor = torch.zeros(container_size, dtype=torch.float32, requires_grad=False).cuda()
+                dist.recv(container_tensor, src, process_group)
+                self.load_from_compacted_states(container_tensor, collected_states, shard_range)
+                del container_tensor
+        else:
+            if own_param:
+                dist.send(compacted_states, master_rank, process_group)
```
Comment (ver217) [inline colossalai/zero/gemini/gemini_optimizer.py:474]: Same problem. Multiple send/recv is inefficient

File: colossalai/zero/gemini/gemini_optimizer.py [hunk 9]
```diff
@@ -301,25 +308,365 @@ def get_range_pair(local_chunk: Chunk, local_param: Parameter):
...
+
+        del compacted_states
+        dist.barrier(process_group)
+
+        # Reshape tensors
+        if rank == master_rank:
+            for state_name, state_tensor in collected_states.items():
+                if state_name == 'step':
+                    continue
+                assert state_tensor.numel() == param.numel()
```
Comment (ver217) [inline colossalai/zero/gemini/gemini_optimizer.py:484]: Can we check if `state_tensor.numel() == param.numel()` first? Hard code about `'step'` is hard to maintain.

File: colossalai/zero/gemini/gemini_optimizer.py (hunk 5 above)
Comment (Fridge003) [reply colossalai/zero/gemini/gemini_optimizer.py:356]: updated

File: colossalai/zero/gemini/gemini_optimizer.py (hunk 7 above)
Comment (Fridge003) [reply colossalai/zero/gemini/gemini_optimizer.py:394]: The master rank might not own the states of a given parameter, in which case states will be empty. So I added get_state_names() method in optimizer classes to fetch names of states.

File: colossalai/zero/gemini/gemini_optimizer.py [hunk 10]
```diff
@@ -301,25 +308,365 @@ def get_range_pair(local_chunk: Chunk, local_param: Parameter):
...
+
+            # Record if master owns this param.
+            if state_shard_range[-1] >= 0:
+                range_info[master_rank] = (state_shard_range[-2].item(), state_shard_range[-1].item())
+
+            container_tensor = torch.zeros_like(state_shard_range, dtype=torch.int, requires_grad=False).cuda()
+            for src in range(0, dist.get_world_size(process_group)):
+                if src == master_rank:
+                    continue
+                dist.recv(container_tensor, src, process_group)
```
[the remark this answers is not in the document]
Comment (Fridge003) [reply colossalai/zero/gemini/gemini_optimizer.py:432]: The states managed by each rank is highly uneven. In most cases, the states of a certain parameter is gathered on one device. The number of parameters whose states are evenly distributed among all devices is small (usually the embedding layer). So I think implementing this feature using all_gather or gather will not help a lot in efficiency.  

'all_gather' might help if we want the complete optimizer states on every rank in the process group, but I think it's not necessary for the implementation of checkpointing.

Also, the frequency of saving to checkpoint is not high, so a little inefficiency doesn't hurt a lot.

File: colossalai/zero/gemini/gemini_optimizer.py (hunk 8 above)
Comment (Fridge003) [reply colossalai/zero/gemini/gemini_optimizer.py:474]: Same response as the last conversation.

File: colossalai/zero/gemini/gemini_optimizer.py (hunk 9 above)
Comment (Fridge003) [reply colossalai/zero/gemini/gemini_optimizer.py:484]: modified

File: colossalai/zero/gemini/gemini_optimizer.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Fridge003) [reply colossalai/zero/gemini/gemini_optimizer.py:435]: I'm still testing its speed, it works well on A100 server. But if it works poorly on weaker GPUs (such as the server for unit testing), I will try to change to all_reduce.

[the remark this answers is not in the document]
Comment (Fridge003) [reply colossalai/zero/gemini/gemini_optimizer.py:435]: OK, I will try to use all_gather

File: colossalai/zero/gemini/gemini_optimizer.py (hunk 7 above)
Comment (ver217) [reply colossalai/zero/gemini/gemini_optimizer.py:404]: But the state is homogeneous for all parameters. You can gather the state names before gathering states.

Comment (ver217) [reply colossalai/zero/gemini/gemini_optimizer.py:404]: Hard coded state names is hard to extend

File: tests/test_checkpoint_io/test_gemini_checkpoint_io.py [hunk 11]
```diff
@@ -29,30 +32,30 @@ def exam_state_dict_with_origin(placement_policy, model_name, use_safetensors: b
...
         new_bert_model = BertForSequenceClassification.from_pretrained(pretrained_path)
-        check_state_dict_equal(bert_model.state_dict(only_rank_0=False, dtype=torch.float32),
+        check_state_dict_equal(bert_model.unwrap().state_dict(only_rank_0=False, dtype=torch.float32),
                                new_bert_model.state_dict(), False)
 
 
+@clear_cache_before_run()
 @parameterize('placement_policy', ['cuda', 'cpu'])
-@parameterize('shard', [True, False])
+@parameterize('shard', [False])
```
Comment (ver217) [inline tests/test_checkpoint_io/test_gemini_checkpoint_io.py:55]: Why disable shard

File: tests/test_checkpoint_io/test_gemini_torch_compability.py [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+    clear_cache_before_run,
+    parameterize,
+    rerun_if_address_is_in_use,
+    spawn,
+)
+from tests.kit.model_zoo import model_zoo
+
+
+@clear_cache_before_run()
+@parameterize('shard', [False])
```
Comment (ver217) [inline tests/test_checkpoint_io/test_gemini_torch_compability.py:22]: Same problem. Why not shard

File: tests/test_checkpoint_io/test_gemini_checkpoint_io.py (hunk 11 above)
Comment (Fridge003) [reply tests/test_checkpoint_io/test_gemini_checkpoint_io.py:55]: True will be added in the following commit of sharded implementation.

File: tests/test_checkpoint_io/test_gemini_torch_compability.py (hunk 12 above)
Comment (Fridge003) [reply tests/test_checkpoint_io/test_gemini_torch_compability.py:22]: True will be added in the following commit of sharded implementation.

File: colossalai/zero/gemini/gemini_optimizer.py (hunk 7 above)
Comment (Fridge003) [reply colossalai/zero/gemini/gemini_optimizer.py:404]: OK, I will try to all_gather state names

Comment (Fridge003) [reply colossalai/zero/gemini/gemini_optimizer.py:422]: Resolved

**Review by ver217 — APPROVED** [review]

**Comment by Fridge003** [comment]
截屏2023-07-07 14 36 13

**Comment by Fridge003** [comment]
截屏2023-07-07 16 29 44

**Pull request merged**